javascript/ql/lib/semmle/javascript/frameworks/Spife.qll

```diff
@@ -303,114 +288,128 @@ module Spife {
   }
 
   /**
-   * A response argument passed to the `reply` method.
+   * A response sent using a method on the `reply` object.
    */
-  private class ReplyArgument extends Http::ResponseSendArgument, DataFlow::Node {
+  private class ReplyMethodCallArgument extends Http::ResponseSendArgument {
+    ReplyCall reply;
+
+    ReplyMethodCallArgument() {
+      // reply.header(RESPONSE, {'Cache-Control': 'no-cache'})
+      reply.getCalleeName() =
+        ["cookie", "link", "header", "headers", "raw", "status", "toStream", "vary"] and
+      reply.getArgument(0) = this
+    }
+
+    override RouteHandler getRouteHandler() { result = reply.getRouteHandler() }
+  }
+
+  /**
+   * A response sent using the `reply()` method.
+   */
+  private class ReplyCallArgument extends Http::ResponseSendArgument {
+    ReplyCall reply;
+
+    ReplyCallArgument() {
+      // reply(RESPONSE, {'Cache-Control': 'no-cache'})
+      reply.isDirectReplyCall() and
+      reply.getArgument(0) = this
+    }
+
+    override RouteHandler getRouteHandler() { result = reply.getRouteHandler() }
+  }
+
+  /**
+   * The return statement for a route handler.
+   */
+  private class RouteHandlerReturn extends Http::ResponseSendArgument {
     RouteHandler rh;
 
-    ReplyArgument() {
-      exists(ReplySource reply, DataFlow::CallNode call |
-        reply.ref() = call and
-        call.getCalleeName() =
-          ["reply", "cookie", "link", "header", "headers", "raw", "status", "toStream", "vary"] and
-        this = call.getArgument(0) and
-        rh = reply.getRouteHandler()
-      )
-      or
-      this = rh.getAReturn()
+    RouteHandlerReturn() {
+      this = rh.getAReturn() and not this.getALocalSource() instanceof ReplyCall
     }
 
     override RouteHandler getRouteHandler() { result = rh }
   }
 
   /**
-   * An expression passed to the `template` method of the reply object
-   * as the value of a template variable.
+   * A call to `reply.template('template', { ... })`, seen as a template instantiation.
    */
-  private class TemplateInput extends Http::ResponseBody {
-    TemplateObjectInput obj;
+  private class TemplateCall extends Templating::TemplateInstantiation::Range instanceof ReplyCall {
+    TemplateCall() { this.getCalleeName() = "template" }
 
-    TemplateInput() {
-      obj.getALocalSource().(DataFlow::ObjectLiteralNode).hasPropertyWrite(_, this)
-    }
+    override DataFlow::SourceNode getOutput() { result = this }
 
-    override RouteHandler getRouteHandler() { result = obj.getRouteHandler() }
+    override DataFlow::Node getTemplateFileNode() { result = this.(ReplyCall).getArgument(0) }
+
+    override DataFlow::Node getTemplateParamsNode() { result = this.(ReplyCall).getArgument(1) }
   }
 
   /**
    * An object passed to the `template` method of the reply object.
    */
   private class TemplateObjectInput extends DataFlow::Node {
-    ReplySource reply;
+    TemplateCall call;
```

Comment:
If you change the type to `ReplyCall`, then you can make the char-pred (and the member predicates) simpler.  

```suggestion
    ReplyCall call;
```